### PR TITLE
MigCluster : Use generateName to create unique secrets

### LINKED ISSUE #396: Clusters and Storage sharing the same name will have secrets fail to create

### PR DESCRIPTION
**Pull request by pranavgaikwad**
Fixes #396 

Proposed Changes :
[1] Creating new MigCluster 
- We use `generateName` instead of `name` in tokenSecret of MigCluster. We also introduce two new labels on the Secret : `createdForResourceType`, `createdForResource`. This makes it easier to find a secret associated with a cluster.
- We create secret first, then get the resulting secret name, then we create MigCluster

[2] Deleting MigCluster
- We find the secret name before deleting MigCluster
- Then usual flow of delete

[3] Editing MigCluster
- Nothing changes here except the secret reference.

### CODE REVIEW

File: src/app/cluster/duck/sagas.ts [hunk 1]
```diff
@@ -304,7 +327,7 @@ function* updateClusterRequest(action) {
 
     // Pushing a request fn to delay the call until its yielded in a batch at same time
     updatePromises.push(() => client.patch(
-      secretResource, clusterValues.name, newTokenSecret));
+      secretResource, currentCluster.spec.serviceAccountSecretRef.name, newTokenSecret));
```
Comment (pranavgaikwad) [inline src/app/cluster/duck/sagas.ts]: Here I am assuming that Cluster name is not changeable from UI

**Comment by pranavgaikwad** [comment]
@ibolton336 added a fix to address the problem you saw. safe guarded the case when - associated Secret does not exist and we try to delete a cluster.

File: src/app/cluster/duck/sagas.ts [hunk 2]
```diff
@@ -132,13 +142,17 @@ function* addClusterRequest(action) {
...
     ]);
 
     const alreadyExists = getResults.reduce((exists, res) => {
-      return res.value && res.value.status === 200 ?
-        [...exists, { kind: res.value.data.kind, name: res.value.data.metadata.name }] :
+      return res && res.status === 200 ?
+        [...exists, { kind: res.value.data.kind, 
+          name:  (res.value.data.items && res.value.data.items.length > 0) 
+            ? res.value.data.items[0].metadata.name : res.value.data.metadata.name
+        }] :
```
Comment (ibolton336) [inline src/app/cluster/duck/sagas.ts]:
```suggestion
      return (res && res.status === 200) ?
        [...exists,
          {
            kind: res.value.data.kind,
            name: (res.value.data.items && res.value.data.items.length > 0) ?
              res.value.data.items[0].metadata.name : res.value.data.metadata.name
          }
        ] :
```

File: src/app/cluster/duck/sagas.ts [hunk 3]
```diff
@@ -87,9 +88,15 @@ function* removeClusterSaga(action) {
       migMeta.configNamespace
     );
     const migClusterResource = new MigResource(MigResourceKind.MigCluster, migMeta.namespace);
+    
+    const secretResourceList = yield client.list(secretResource, 
+      getTokenSecretLabelSelector(MigResourceKind.MigCluster, name));
+
+    const secretResourceName = (secretResourceList.data.items && secretResourceList.data.items.length > 0) 
+      ? secretResourceList.data.items[0].metadata.name : '';
```
Comment (ibolton336) [inline src/app/cluster/duck/sagas.ts]:
```suggestion
    const secretResourceName = (secretResourceList.data.items && secretResourceList.data.items.length > 0) ?
      secretResourceList.data.items[0].metadata.name : '';
```

**Pull request merged**